From fba80fe8421c23bdc0e1403d4694a2b008c5aa51 Mon Sep 17 00:00:00 2001
From: Dana Fried <dfried@chromium.org>
Date: Wed, 21 Jan 2026 06:14:27 -0800
Subject: [PATCH] [Vertical Tabs] Paint tabstrip to layer

This prevents the animating side panel from painting over the tabstrip,
since Z-order works like:

layer highest Z
...
layer lowest Z
non-layer highest Z
...
non-layer lowest Z

Bug: 473424028
Change-Id: I09e4f2c60c5dd45583c15523316a940b85ecab35
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/7494682
Commit-Queue: Dana Fried <dfried@chromium.org>
Auto-Submit: Dana Fried <dfried@chromium.org>
Reviewed-by: Eshwar Stalin <estalin@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1572266}
---

--- a/chrome/browser/ui/views/frame/vertical_tab_strip_region_view.cc
+++ b/chrome/browser/ui/views/frame/vertical_tab_strip_region_view.cc
@@ -37,6 +37,7 @@
 #include "components/tabs/public/tab_interface.h"
 #include "ui/base/metadata/metadata_impl_macros.h"
 #include "ui/color/color_id.h"
+#include "ui/compositor/layer.h"
 #include "ui/gfx/animation/animation.h"
 #include "ui/views/background.h"
 #include "ui/views/controls/resize_area.h"
@@ -61,6 +62,11 @@ VerticalTabStripRegionView::VerticalTabS
     : tab_strip_model_(browser->GetTabStripModel()),
       state_controller_(state_controller),
       resize_animation_(this) {
+  // For z-ordering purposes this needs to be on a layer.
+  SetPaintToLayer();
+  // Because corners may be transparent, this must be set to false.
+  layer()->SetFillsBoundsOpaquely(false);
+
   flex_layout_ = SetLayoutManager(std::make_unique<views::FlexLayout>());
   flex_layout_->SetOrientation(views::LayoutOrientation::kVertical)
       .SetCollapseMargins(true)
@@ -471,6 +477,8 @@ void VerticalTabStripRegionView::UpdateB
   top_button_separator_->SetColorId(IsFrameActive()
                                         ? kColorTabDividerFrameActive
                                         : kColorTabDividerFrameInactive);
+
+  SchedulePaint();
 }
 
 bool VerticalTabStripRegionView::IsFrameActive() const {
